credentials/credentials.go

```diff
@@ -126,6 +126,10 @@ func (t TLSInfo) AuthType() string {
 	return "tls"
 }
 
+func (t TLSInfo) ConnState() tls.ConnectionState {
```

Comment:
On Tue, Sep 8, 2015 at 11:00 AM, Tamir Duberstein <notifications@github.com>
wrote:

> In credentials/credentials.go
> <https://github.com/grpc/grpc-go/pull/328#discussion_r38958528>:
>
> > @@ -126,6 +126,10 @@ func (t TLSInfo) AuthType() string {
> >  	return "tls"
> >  }
> >
> > +func (t TLSInfo) ConnState() tls.ConnectionState {
>
> Making it exported also makes a copy when users work on it (notice
> tls.ConnectionState is a object not a pointer).
>
> Doesn't matter that it's not a pointer. If it's exported I can overwrite
> it entirely: tlsInfo.State = myOtherTlsState.
>
I would say overwriting does not matter because you only overwrite your own
copy which is not the ConnectionState of the real tls connection (because
tls.ConnectionState returns a copy). I think that is the whole point why
tls.ConnectionState returns an object instead of a pointer. Keep in mind
that you are the service provider. I do not see any malicious operations
could happen here.

> Plus, making a copy does
> NOT prohibit mutation because tls.ConnectionState contains a number of
> pointer fields.
>
> Yes, you could mess with some of the slices in tls.ConnectionState
> <http://golang.org/pkg/crypto/tls/#ConnectionState> but it's still better
> than exporting this.
>
Slice elements are pointers.

> Actually adding a getter might make the 2nd copy (depends on how the
> compiler optimizes it) which is not needed.
>
> This doesn't matter - those copies are on the stack and the Go compiler
> does RVO.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/328/files#r38958528>.
>